library/kani_macros/src/sysroot/contracts/mod.rs

```diff
@@ -180,56 +156,106 @@
 //! ```
 //!
 //! Turns into
-//!
 //! ```
-//! #[kanitool::checked_with = "div_recursion_wrapper_965916"]
-//! #[kanitool::replaced_with = "div_replace_965916"]
-//! fn div(dividend: u32, divisor: u32) -> u32 { dividend / divisor }
-//!
-//! #[allow(dead_code, unused_variables)]
-//! #[kanitool :: is_contract_generated(check)] fn
-//! div_check_b97df2(dividend : u32, divisor : u32) -> u32
-//! {
-//!     let dividend_renamed = kani::internal::untracked_deref(& dividend);
-//!     let divisor_renamed = kani::internal::untracked_deref(& divisor);
-//!     kani::assume(divisor != 0);
-//!     let result_kani_internal : u32 = div_wrapper_b97df2(dividend, divisor);
-//!     kani::assert(
-//!     (| result : & u32 | *result <= dividend_renamed)(& result_kani_internal),
-//!     stringify!(|result : &u32| *result <= dividend));
-//!     std::mem::forget(dividend_renamed);
-//!     std::mem::forget(divisor_renamed);
-//!     result_kani_internal
-//! }
-//!
-//! #[allow(dead_code, unused_variables)]
-//! #[kanitool :: is_contract_generated(replace)] fn
-//! div_replace_b97df2(dividend : u32, divisor : u32) -> u32
-//! {
-//!     let divisor_renamed = kani::internal::untracked_deref(& divisor);
-//!     let dividend_renamed = kani::internal::untracked_deref(& dividend);
-//!     kani::assert(divisor != 0, stringify! (divisor != 0));
-//!     let result_kani_internal : u32 = kani::any_modifies();
-//!     kani::assume(
-//!     (|result : & u32| *result <= dividend_renamed)(&result_kani_internal));
-//!     std::mem::forget(divisor_renamed);
-//!     std::mem::forget(dividend_renamed);
-//!     result_kani_internal
-//! }
-//!
-//! #[allow(dead_code)]
-//! #[allow(unused_variables)]
-//! #[kanitool::is_contract_generated(recursion_wrapper)]
-//! fn div_recursion_wrapper_965916(dividend: u32, divisor: u32) -> u32 {
-//!     static mut REENTRY: bool = false;
-//!
-//!     if unsafe { REENTRY } {
-//!         div_replace_b97df2(dividend, divisor)
-//!     } else {
-//!         unsafe { reentry = true };
-//!         let result_kani_internal = div_check_b97df2(dividend, divisor);
-//!         unsafe { reentry = false };
-//!         result_kani_internal
+//! #[kanitool::recursion_check = "__kani_recursion_check_div"]
+//! #[kanitool::checked_with = "__kani_check_div"]
+//! #[kanitool::replaced_with = "__kani_replace_div"]
+//! #[kanitool::inner_check = "__kani_modifies_div"]
+//! fn div(dividend: u32, divisor: u32) -> u32 {
+//!     #[inline(never)]
+//!     #[kanitool::fn_marker = "kani_register_contract"]
+//!     pub const fn kani_register_contract<T, F: FnOnce() -> T>(f: F) -> T {
+//!         kani::panic("internal error: entered unreachable code: ")
+//!     }
```

Comment:
Why is this register necessary? Couldn't we just get the mode, set the `kani_contract_mode` and call the original function?

Reply:
Talked offline, these are needed to allow contract of constant functions to contain non-constant logic.